bigchaindb/voting.py

```diff
@@ -0,0 +1,152 @@
+import collections
+
+from bigchaindb.common.schema import SchemaValidationError, validate_vote_schema
+from bigchaindb.common.utils import serialize
+from bigchaindb.common.crypto import PublicKey
+
+
+VALID = 'valid'
+INVALID = 'invalid'
+UNDECIDED = 'undecided'
+
+
+class Voting:
+    """
+    Everything to do with creating and checking votes.
+
+    All functions in this class should be referentially transparent, that is,
+    they always give the same output for a given input. This makes it easier
+    to test. This also means no logging!
+
+    Assumptions regarding data:
+        * Vote is a dictionary, but it is not assumed that any properties are.
+        * Everything else is assumed to be structurally correct, otherwise errors
+        may be thrown.
+    """
+
+    @classmethod
+    def block_election(cls, block, votes, keyring):
+        """
+        Calculate the election status of a block.
+        """
+        eligible_voters = set(block['block']['voters']) & set(keyring)
+        eligible_votes, ineligible_votes = \
+            cls.partition_eligible_votes(votes, eligible_voters)
+        n_voters = len(eligible_voters)
+        results = cls.count_votes(eligible_votes)
+        results['status'] = cls.decide_votes(n_voters, **results['counts'])
+        results['ineligible'] = ineligible_votes
+        return results
+
+    @classmethod
+    def partition_eligible_votes(cls, votes, eligible_voters):
+        """
+        Filter votes from unknown nodes or nodes that are not listed on
```

Comment:
Well yes; with the exception that, if there is a badly constructed / wrong previous block vote but correctly signed, it counts towards the INVALID count. This reduces the chances that a block gets stuck as UNDECIDED.